migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java

```diff
@@ -44,61 +48,127 @@
+    ImmutableMap<Column<?>, Object> recipeRow = getRecipesCsvTable().getRow(recipeKey);
+
+    ImmutableList.Builder<MigrationResult> migratedResults = ImmutableList.builder();
+    if (!recipeRow.get(RecipesTableSchema.REGION_CODE).equals(regionCode)) {
+      return migratedResults.build();
+    }
+    migratableRange.forEach(entry -> {
+        migratedResults.add(migrate(entry.getSanitizedNumber(), recipeRow, entry.getOriginalNumber()));
+    });
 
-    return getNumberRange()
-        .filter(regionalRecipes::contains);
+    // TODO: create MigrationReport class holding details of migration and return it here instead
+    return migratedResults.build();
   }
 
   /**
-   * Returns the sub range of numbers within numberRange that can be migrated using the given
-   * recipe. This method will not perform migrations and as a result, the validity of migrations
-   * using the given recipe cannot be verified.
+   * Takes a given number and migrates it using the given matching recipe row. If the given recipe
+   * is not a final migration, the method is recursively called with the recipe that matches the new
+   * migrated number until a recipe that produces a final migration (a recipe that results in the
+   * new format being valid and dialable) has been used. Once this occurs, the {@link MigrationResult}
+   * is returned.
    *
-   * @param recipeKey: the key of the recipe that is being checked
-   * @throws IllegalArgumentException if there is no row in the recipesTable with the given
-   * recipeKey
+   * @throws IllegalArgumentException if the 'Old Format' value in the given recipe row does not
+   * match the number to migrate. This means that the 'Old Format' value cannot be represented by
+   * the given recipes 'Old Prefix' and 'Old Length'.
+   * @throws RuntimeException when the given recipe is not a final migration and a recipe cannot be
+   * found in the recipesTable to match the resulting number from the initial migrating recipe.
    */
-  public Stream<DigitSequence> getMigratableNumbers(RangeKey recipeKey) {
-    if (!recipesTable.containsRow(recipeKey)) {
+  private MigrationResult migrate(DigitSequence migratingNumber,
+      ImmutableMap<Column<?>, Object> recipeRow,
+      String originalNumber) {
+    String oldFormat = (String) recipeRow.get(RecipesTableSchema.OLD_FORMAT);
+    String newFormat = (String) recipeRow.get(RecipesTableSchema.NEW_FORMAT);
+
+    if (!RangeSpecification.parse(oldFormat).matches(migratingNumber)) {
       throw new IllegalArgumentException(
-          recipeKey + " does not match any recipe row in the given recipes table");
+          "value '" + oldFormat + "' in column 'Old Format' cannot be"
+              + " represented by its given recipe key (Old Prefix + Old Length)");
+    }
+    String newString = getNewString(migratingNumber.toString(), oldFormat, newFormat);
+
+    if (!(boolean) recipeRow.get(RecipesTableSchema.IS_FINAL_MIGRATION)) {
```

Comment:
Optional: Or

x != Boolean.TRUE - safe but IDE might complain
!Boolean.TRUE.equals(x) - safe

boolean -> Boolean (auto-boxing, safe)
Boolean -> boolean (unboxing, not 100% safe if null, gives null pointer exception)